src/java/org/apache/cassandra/index/sai/disk/v1/bbtree/BlockBalancedTreeWriter.java

```diff
@@ -0,0 +1,779 @@
+                    assert lastLevel >= 0;
+                    if (lastLevel != 0)
+                    {
+                        // Last level is partially filled, so we must rotate the leaf FPs to match.  We do this here, after loading
+                        // at read-time, so that we can still delta code them on disk at write:
+                        long[] newLeafBlockFPs = new long[numLeaves];
+                        System.arraycopy(leafBlockFPs, lastLevel, newLeafBlockFPs, 0, leafBlockFPs.length - lastLevel);
+                        System.arraycopy(leafBlockFPs, 0, newLeafBlockFPs, leafBlockFPs.length - lastLevel, lastLevel);
+                        leafBlockFPs = newLeafBlockFPs;
+                    }
+                    break;
+                }
+
+                levelCount *= 2;
+            }
+        }
+
+        // Reused while packing the index
+        try (RAMIndexOutput writeBuffer = new RAMIndexOutput("PackedIndex"))
+        {
+            // This is the "file" we append the byte[] to:
+            List<byte[]> blocks = new ArrayList<>();
+            byte[] lastSplitValues = new byte[bytesPerValue];
+            int totalSize = recursePackIndex(writeBuffer, leafBlockFPs, splitPackedValues, 0, blocks, 1, lastSplitValues, new boolean[1], false);
+            // Compact the byte[] blocks into single byte index:
+            byte[] index = new byte[totalSize];
+            int upto = 0;
+            for (byte[] block : blocks)
+            {
+                System.arraycopy(block, 0, index, upto, block.length);
+                upto += block.length;
+            }
+            assert upto == totalSize;
+
+            return index;
+        }
+    }
+
+    /**
+     * lastSplitValues is per-dimension split value previously seen; we use this to prefix-code the split byte[] on each
+     * inner node
+     */
+    private int recursePackIndex(RAMIndexOutput writeBuffer, long[] leafBlockFPs, byte[] splitPackedValues, long minBlockFP, List<byte[]> blocks,
+                                 int nodeID, byte[] lastSplitValues, boolean[] negativeDeltas, boolean isLeft) throws IOException
+    {
+        if (nodeID >= leafBlockFPs.length)
+        {
+            int leafID = nodeID - leafBlockFPs.length;
+
+            // In the unbalanced case it's possible the left most node only has one child:
+            if (leafID < leafBlockFPs.length)
+            {
+                long delta = leafBlockFPs[leafID] - minBlockFP;
+                if (isLeft)
+                {
+                    assert delta == 0;
+                    return 0;
+                }
+                else
+                {
+                    assert nodeID == 1 || delta > 0 : "nodeID=" + nodeID;
+                    writeBuffer.writeVLong(delta);
+                    return appendBlock(writeBuffer, blocks);
+                }
+            }
+            else
+            {
+                return 0;
```

Comment:
nit: Untested code path?

Reply:
Yes, and I have tried to reach this point in testing. I will keep looking at this but it I have yet to determine what it means by an unbalanced tree.